src/libfetchers/git.cc

```diff
@@ -225,17 +225,25 @@ struct GitInputScheme : InputScheme
 
             /* Check whether HEAD points to something that looks like a commit,
                since that is the refrence we want to use later on. */
-            bool hasHead = false;
-            try {
-                runProgram("git", true, { "-C", actualUrl, "rev-parse", "--verify", "--no-revs", "HEAD^{commit}" });
-                hasHead = true;
-            } catch (ExecError & e) {
-                // git exits with status 128 here if it does not detect a repository.
-                if (!WIFEXITED(e.status) || WEXITSTATUS(e.status) != 128) {
-                    throw Error("Git tree '%s' is broken.\n'git rev-parse --verify HEAD' failed with exit code %d.", actualUrl, WEXITSTATUS(e.status));
-                }
+            auto result = runProgram(RunOptions {
+                .program = "git",
+                .args = { "-C", actualUrl, "--git-dir=.git", "rev-parse", "--verify", "--no-revs", "HEAD^{commit}" },
+                .mergeStderrToStdout = true
+            });
+            auto exitCode = WEXITSTATUS(result.first);
+            auto errorMessage = result.second;
+
+            if (errorMessage.find("fatal: not a git repository") != std::string::npos) {
+                throw Error("'%s' is not a git repository.", actualUrl);
+            } else if (errorMessage.find("fatal: Needed a single revision") != std::string::npos) {
```

Comment:
That’s not gonna work if the locale isn’t english because the error message will be translated:

```console
$ git init
$ nix flake init                                                 
$ LANG=C nix run github:mschwaig/nix/fix-git-workspace-dirty-detection -- build # Works with a C (or en) locale
warning: Git tree '/tmp/tmp.hjV72WecGo' is dirty
warning: creating lock file '/tmp/tmp.hjV72WecGo/flake.lock'
warning: Git tree '/tmp/tmp.hjV72WecGo' is dirty
$ LANG=fr_FR nix run github:mschwaig/nix/fix-git-workspace-dirty-detection -- build # But fails with another
error: Getting the HEAD of the git tree '/tmp/tmp.hjV72WecGo' failed with exit code 128:
       fatal: Une seule rision attendue
       
(use '--show-trace' to show detailed location information)
```

Maybe you could run another command here (`git rev-parse --all` ? `git rev-parse --git-dir` ? might be a better one) to distinguish between the no-repo and repo-but-no-commit cases. Since it’s an edge-case the cost of an extra command shouldn’t be an issue

Reply:
I have implemented a fix for this now in 0ef6df466f46a32f8c8677009303aea7d98ea2ee.

I have opted to explicitly setting the locale via an environment variable and still parsing the error messages, instead of running another `git` command, because I do not only want to distinguish the  no-repo and repo-but-no-commit cases, but also the case where `git` fails in some other unexpected error, which  should lead to fetch failing. I think looking at the error messages works better for that.